File: eval/holdout-unity/t2_net/Assets/Tests/Holdout/HoldoutNet.cs
// HELD-OUT. The agent never sees this file.
// Grades the "draw a centre net" task by looking at real rendered pixels.
using NUnit.Framework;
using Pong.Sim;
using UnityEngine;
using Pong.View;

namespace Pong.Holdout
{
    public class HoldoutNet
    {
        static readonly Color32 Bg = ViewConfig.BackgroundRgb;
        const int Tol = 8;

        static Frame CaptureOrSkip(ulong seed, int ticks)
        {
            try { return RenderHarness.CaptureFrame(seed, ticks, null); }
            catch (System.Exception e) { Assert.Ignore($"no usable graphics device: {e.Message}"); return null; }
        }

        static bool IsInk(Frame f, int x, int y)
        {
            var p = f.Pixel(x, y);
            return System.Math.Abs(p.r - Bg.r) > Tol
                || System.Math.Abs(p.g - Bg.g) > Tol
                || System.Math.Abs(p.b - Bg.b) > Tol;
        }

        [Test]
        public void ACentreNetIsDrawnDownTheMiddle()
        {
            var frame = CaptureOrSkip(11, 20);
            int mid = frame.Width / 2, litRows = 0;
            for (int y = 0; y < frame.Height; y++)
                for (int x = mid - 3; x <= mid + 3; x++)
                    if (IsInk(frame, x, y)) { litRows++; break; }

            float fraction = (float)litRows / frame.Height;
            Assert.Greater(fraction, 0.30f,
                $"expected a visible net: only {litRows}/{frame.Height} rows " +
                $"({fraction * 100:F0}%) have ink within 3px of the middle. A ball " +
                "alone lights a few rows; a net lights most of them.");
        }

        [Test]
        public void TheNetDoesNotCoverThePlayArea()
        {
            var frame = CaptureOrSkip(12, 20);
            float coverage = frame.InkCoverage(Bg, Tol);
            Assert.Less(coverage, 0.25f,
                $"{coverage * 100:F1}% of the frame is non-background - the net is far too wide");
        }

        [Test]
        public void AddingTheNetDidNotBreakThePaddles()
        {
            var frame = CaptureOrSkip(13, 20);
            bool BandHasInk(int lo, int hi)
            {
                for (int x = lo; x < hi; x++)
                    for (int y = 0; y < frame.Height; y++)
                        if (IsInk(frame, x, y)) return true;
                return false;
            }
            Assert.IsTrue(BandHasInk(0, frame.Width / 6), "left paddle disappeared");
            Assert.IsTrue(BandHasInk(frame.Width * 5 / 6, frame.Width), "right paddle disappeared");
        }
    }
}

File: eval/holdout/t2_net/crates/game/tests/holdout_net.rs
//! HELD-OUT. The agent never sees this file.
//!
//! Grades the "draw a centre net" task by looking at actual rendered pixels.

use game::harness::capture_frame;
use game::BACKGROUND_COLOR;

fn background_rgb() -> [u8; 3] {
    let s = BACKGROUND_COLOR.to_srgba();
    [
        (s.red * 255.0).round() as u8,
        (s.green * 255.0).round() as u8,
        (s.blue * 255.0).round() as u8,
    ]
}

#[test]
fn a_centre_net_is_drawn_down_the_middle() {
    let frame = match capture_frame(11, 20, &[]) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("SKIP: no GPU ({e})");
            return;
        }
    };
    let bg = background_rgb();
    let mid = frame.width / 2;

    // Count rows that have ink somewhere in a narrow band at the exact centre.
    // The ball is small and only occupies a handful of rows, so a passing score
    // here genuinely requires a full-height net.
    let lit_rows = (0..frame.height)
        .filter(|&y| {
            ((mid - 3)..=(mid + 3)).any(|x| {
                let p = frame.pixel(x, y);
                (0..3).any(|c| p[c].abs_diff(bg[c]) > 8)
            })
        })
        .count();

    let fraction = lit_rows as f32 / frame.height as f32;
    assert!(
        fraction > 0.30,
        "expected a visible net down the centre: only {lit_rows}/{} rows \
         ({:.0}%) have ink within 3px of the middle. A ball alone lights a few \
         rows; a net lights most of them.",
        frame.height,
        fraction * 100.0
    );
}

#[test]
fn the_net_does_not_cover_the_play_area() {
    let frame = match capture_frame(12, 20, &[]) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("SKIP: no GPU ({e})");
            return;
        }
    };
    let coverage = frame.ink_coverage(background_rgb(), 8);
    assert!(
        coverage < 0.25,
        "{:.1}% of the frame is non-background - the net is far too wide or \
         something is covering the arena",
        coverage * 100.0
    );
}

#[test]
fn adding_the_net_did_not_break_the_paddles() {
    let frame = match capture_frame(13, 20, &[]) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("SKIP: no GPU ({e})");
            return;
        }
    };
    let bg = background_rgb();
    let band_has_ink = |lo: u32, hi: u32| {
        (lo..hi).any(|x| {
            (0..frame.height).any(|y| {
                let p = frame.pixel(x, y);
                (0..3).any(|c| p[c].abs_diff(bg[c]) > 8)
            })
        })
    };
    assert!(band_has_ink(0, frame.width / 6), "left paddle disappeared");
    assert!(
        band_has_ink(frame.width * 5 / 6, frame.width),
        "right paddle disappeared"
    );
}
